### ISSUE TITLE
Perf: potential lock contention in ProjectRootElementCache.Get

### ISSUE THREAD

**Issue by lifengl** [dotnet/msbuild#3039]
When we look into performance traces of parallel project loading work, we noticed some lock contentions in ProjectRootElementCache.Get method.  In one trace we noticed it was blocked 60ms.

The code is using a single lock to control to access the cache, and also read file into XmlDocument, when it does not exist.  The later one can be slow on a slow disk.  When PPL is enabled, multiple projects are evaluated, and this becomes more a problem. 

If the code can pull out the logic to load a project file out of the lock region, it will allow other threads to access the cache for other files in the same project collection.

**Comment by lifengl** [comment]
When Jialong tries to open a solution with 1600 projects, this lock contention issue became much more serious, and the main reason to slow down the phase of loading construction model of projects during the solution loading time. (During solution loading time, CPS loads the construction model in one large batch, before it starts to evaluate them). 

Name                                                 	Inc %	     Inc	Exc %	   Exc	Fold	                             When	     First	       Last
 clr!AwareLock::Contention                           	  9.0	   9,533	  8.8	 9,341	 564	 __03.4444442._o___.ooo..ooooo_.o	 6,030.473	 77,354.601
+ clr!JITutil_MonContention                          	  8.9	   9,527	  0.0	     0	   0	 __0354444442__o___.ooo..ooo.o_.o	 6,030.473	 77,354.601
|+ microsoft.build.ni!ProjectRootElementCache.Get    	  8.9	   9,455	  0.0	     0	   0	 __0o54444442__o___.______o_.o__.	 6,030.473	 76,220.207
||+ microsoft.build.ni!ProjectRootElement.Open       	  7.5	   7,972	  0.0	     0	   0	 __0o43443332__o___.______o_.o__.	 6,030.473	 76,220.207
||+ microsoft.build.ni!ProjectRootElementCache.TryGet	  1.4	   1,483	  0.0	     0	   0	 __o000000000____________________	 6,893.559	 28,217.690

**Comment by davkean** [comment]
Just the overhead of taking the lock itself (not the blocked time) is ~0.9% CPU opening a large solution of ~1400 projects. 

[Trace](https://microsoft.sharepoint.com/:f:/t/VSIDEPerformance/Etm7IDAYTSpIn9CAQYfXgYUBuUMciQ4xUku3e1KWZtLspw?e=YOQ2sB) (1404 projects)

```
Name | Inc % | Inc
-- | -- | --
clr!AwareLock::Contention | 1.0 | 5,491
+ clr!JITutil_MonContention | 1.0 | 5,488
\|+ microsoft.build.ni!ProjectRootElementCache.Get | 0.9 | 4,710
```

**Comment by AR-May** [comment]
Status update: issue is in progress, I am actively working on a PR for it. 
ETA for the PR in the MSBuild repo: end of week (in two days).

**Comment by AR-May** [comment]
Status Update: Fixed in #6680. As suggested, we pulled out the logic of loading a project file out of the single lock region. We decided against locking the loading region with any project specific lock, as we are fine with possibility of loading the same file twice, as this would be quite rare situation.
